File: JS/1507.reformat_date [2].js
/**
 * @param {string} date
 * @return {string}
 */

function getMonthNum(monthStr){
    const months = [
        {name: 'Jan', val: '01'},
        {name: 'Feb', val: '02'},
        {name: 'Mar', val: '03'},
        {name: 'Apr', val: '04'},
        {name: 'May', val: '05'},
        {name: 'Jun', val: '06'},
        {name: 'Jul', val: '07'},
        {name: 'Aug', val: '08'},
        {name: 'Sep', val: '09'},
        {name: 'Oct', val: '10'},
        {name: 'Nov', val: '11'},
        {name: 'Dec', val: '12'}
    ]

    return months.find(({name}) => name === monthStr).val
}

var reformatDate = function(date) {
    const len = date.length
    
    const dd = len == 12 ? '0' + date.substring(0, 1) : date.substring(0, 2)
    const mm = len == 12 ? getMonthNum(date.substring(4, 7)) : getMonthNum(date.substring(5, 8))
    const yyyy = len == 12 ? date.substring(8) : date.substring(9)

    return yyyy + '-' + mm + '-' + dd
};

File: CPP [1501-2000]/1507.reformat-date [2].cpp
class Solution {
public:
    string getMonthNum(string month){
        vector<string> months = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};

        string monthNum = "";
        for(int i=0; i<12; i++)
            if(month == months[i])
            {
                monthNum += to_string(i+1);
                break;
            }
            
        return monthNum.length() == 1 ? '0' + monthNum : monthNum;
    }

    string reformatDate(string date) {
        int len = date.length();

        string ddd = len == 12 ? '0' + date.substr(0, 1) : date.substr(0, 2);
        string mm = len == 12 ? getMonthNum(date.substr(4, 3)) : getMonthNum(date.substr(5, 3));
        string yyyy = len == 12 ? date.substr(8) : date.substr(9);

        return yyyy + "-" + mm + "-" + ddd;
    }
};
/*
# date.length() can be only 12 or 13
# 12 when the date is single-digit like "6th Jun 1933"
# 13 when the date is double-digit like "20th Oct 2052"
*/